src/logics/Logic.cc

```diff
@@ -1600,3 +1607,108 @@ char* Logic::printTerm        (PTRef tr, bool l, bool s) const { return printTer
 void Logic::termSort(vec<PTRef>& v) const { sort(v, LessThan_PTRef()); }
 
 void  Logic::purify     (PTRef r, PTRef& p, lbool& sgn) const {p = r; sgn = l_True; while (isNot(p)) { sgn = sgn^1; p = getPterm(p)[0]; }}
+
+bool Logic::typeCheck(SymRef sym, vec<PTRef> const & args, std::string & why) const {
+
+    auto genSortMismatchString = [&](SymRef sym, vec<PTRef> const & args) {
+        std::string symStr = getSymName(sym);
+        Symbol const & symbol = sym_store[sym];
+        if (symbol.chainable() or symbol.pairwise()) {
+            for (auto _ : args) {
+                (void)_;
+                symStr += " " + std::string(getSortName(symbol[0]));
+            }
+        } else if (symbol.left_assoc()) {
+            symStr += " " + std::string(getSortName(symbol[0]));
+            for (auto _ : opensmt::span(args.begin() + 1, args.size()-1)) {
+                (void)_;
+                symStr += " " + std::string(getSortName(symbol[1]));
+            }
+        } else if (symbol.right_assoc()) {
+            for (auto _ : opensmt::span(args.begin(), args.size()-1)) {
+                (void)_;
+                symStr += " " + std::string(getSortName(symbol[0]));
+            }
+            symStr += " " + std::string(getSortName(symbol[1]));
+        }
+
+        std::string argSorts;
+        for (PTRef tr : args) {
+            argSorts += std::string(getSortName(getSortRef(tr))) + " ";
+        }
+        return "Symbol " + symStr + " instantiated with arguments of sort " + argSorts;
+    };
+
+    auto genArgNumMismatchString = [&](SymRef sym, int expectedArgs, int actualArgs) {
+        return "Symbol " + std::string(getSymName(sym)) + " expects " + std::to_string(expectedArgs) +
+        " but " + std::to_string(actualArgs) + " were provided. ";
+    };
+
+    Symbol const & symbol = sym_store[sym];
+
+    if (symbol.chainable() or symbol.pairwise()) {
+        // Need to have at least two arguments and they all need to be of the same sort
+        if (args.size() < 2) {
+            why.assign(genArgNumMismatchString(sym, 2, args.size()));
+            return false;
+        }
+        SRef argSort = getSortRef(args[0]);
+        if (argSort != symbol[0]) {
+            why.assign(genSortMismatchString(sym, args));
+            return false;
+        }
+        if (not std::all_of(args.begin(), args.end(), [&](PTRef tr) { return argSort == getSortRef(tr); })) {
+            why.assign(genSortMismatchString(sym, args));
+            return false;
+        }
```

Comment:
These two ifs have the same body, can we merge them?

Reply:
fixed